File: simple_loader.php
<?php

class SimpleLoader {

    /**
     * Displays a progress bar.
     *
     * @param int $totalSteps Total steps for completion.
     * @param string $prefix Prefix string.
     * @param string $suffix Suffix string.
     * @param int $length Length of the bar.
     * @param string $fill Bar fill character.
     */
    public static function loadingBar($totalSteps, $prefix = 'Loading', $suffix = 'Complete', $length = 50, $fill = '█') {
        for ($step = 0; $step <= $totalSteps; $step++) {
            $percent = ($step / $totalSteps) * 100;
            $filledLength = floor($length * $step / $totalSteps);
            $bar = str_repeat($fill, $filledLength) . str_repeat('-', $length - $filledLength);
            echo "\r{$prefix} |{$bar}| " . number_format($percent, 1) . "% {$suffix}";
            usleep(100000); // Simulate work with delay (100ms)
        }
        echo "\n"; // New line after completion
    }

    /**
     * Displays a spinning loader.
     *
     * @param int $duration Duration in seconds.
     * @param string $message Message to display alongside the spinner.
     * @param string $spinnerChars Characters to use for the spinner.
     */
    public static function spinningLoader($duration = 5, $message = 'Processing', $spinnerChars = '|/-\\') {
        $stopSpinner = false;
        $spinnerChars = str_split($spinnerChars);
        $i = 0;

        // Start the spinner in a loop
        $spinnerThread = function () use (&$stopSpinner, $message, $spinnerChars, &$i) {
            while (!$stopSpinner) {
                echo "\r{$message} {$spinnerChars[$i]}";
                flush();
                $i = ($i + 1) % count($spinnerChars);
                usleep(100000); // Sleep for 100ms between each character
            }
            echo "\rDone!       ";
        };

        // Run the spinner in the background
        $spinnerThread();

        // Simulate work for the given duration
        sleep($duration);

        // Stop the spinner after the duration
        $stopSpinner = true;
    }
}

// Loading bar example
SimpleLoader::loadingBar(100, 'Downloading', 'Complete', 50);

// Spinning loader example
SimpleLoader::spinningLoader(5, 'Processing', '|/-\\');

?>

File: simple_loader.rs
use std::{thread, time};

pub struct SimpleLoader;

impl SimpleLoader {
    /// Displays a progress bar.
    pub fn loading_bar(total_steps: usize, prefix: &str, suffix: &str, length: usize, fill: char) {
        let print_bar = |step: usize| {
            let percent = (step as f64 / total_steps as f64) * 100.0;
            let filled_length = (length * step) / total_steps;
            let bar: String = fill.to_string().repeat(filled_length) + &"-".repeat(length - filled_length);
            print!("\r{} |{}| {:.1}% {}", prefix, bar, percent, suffix);
            std::io::stdout().flush().unwrap();
        };

        for step in 0..=total_steps {
            print_bar(step);
            thread::sleep(time::Duration::from_millis(100)); // Simulate work
        }
        println!(); // New line after completion
    }

    /// Displays a spinning loader.
    pub fn spinning_loader(duration: u64, message: &str, spinner_chars: &str) {
        let mut stop_spinner = false;
        let spinner_chars = spinner_chars.chars().collect::<Vec<char>>();
        let mut idx = 0;

        thread::spawn(move || {
            while !stop_spinner {
                print!("\r{} {}", message, spinner_chars[idx]);
                std::io::stdout().flush().unwrap();
                idx = (idx + 1) % spinner_chars.len();
                thread::sleep(time::Duration::from_millis(100));
            }
            print!("\rDone!       ");
        });

        thread::sleep(time::Duration::from_secs(duration)); // Simulate duration
        stop_spinner = true;
    }
}

fn main() {
    // Loading bar example
    SimpleLoader::loading_bar(100, "Downloading", "Complete", 50, '█');
    
    // Spinning loader example
    SimpleLoader::spinning_loader(5, "Processing", "|/-\\");
}
